subsys/bluetooth/host/scan.c

```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
-			 * discard the partially received report, and the application is not
-			 * notified.
-			 *
-			 * See the Controller's documentation for possible reasons for aborting.
-			 * Hint: CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX.
-			 */
-			LOG_DBG("Discarding incomplete advertisement.");
-			reset_reassembling_advertiser();
-			goto cont;
-		}
+	bool complete = data_status == COMPLETE;
+	bool borked = (data_status == INCOMPLETE_DATA_TRUNCATED_NO_MORE_TO_COME) ||
+		      (data_status == CONTACT_SALES);
+	bool end_of_chain = complete || borked;
 
-		if (is_new_advertiser) {
-			/* We are not reassembling reports from an advertiser and
-			 * this is the first report from the new advertiser.
-			 * Initialize the new advertiser.
-			 */
-			__ASSERT_NO_MSG(reassembling_advertiser.state == FRAG_ADV_INACTIVE);
-			init_reassembling_advertiser(&evt->addr, evt->sid);
-		}
+	bool reassembling = r->state != FRAG_ADV_INACTIVE;
+	bool same_chain = advertisers_equal(r, &evt->addr, evt->sid);
 
-		if (evt->length + ext_scan_buf.len > ext_scan_buf.size) {
-			/* The report does not fit in the reassemby buffer
-			 * Discard this and future reports from the advertiser.
-			 */
-			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
-		}
+	LOG_DBG("[%s] status %s", bt_addr_le_str(&evt->addr), status_to_str(data_status));
 
-		if (reassembling_advertiser.state == FRAG_ADV_DISCARDING) {
-			if (!more_to_come) {
-				/* We do no longer need to keep track of this advertiser as
-				 * all the expected data is received.
-				 */
-				reset_reassembling_advertiser();
-			}
-			goto cont;
-		}
+	if (complete && !same_chain) {
+		/* Interleaved "complete" reports are allowed (for some reason). */
+		return false;
+	}
 
-		net_buf_simple_add_mem(&ext_scan_buf, buf->data, evt->length);
-		if (more_to_come) {
-			/* The controller will send additional reports to be reassembled */
-			continue;
-		}
+	if (reassembling && !same_chain) {
+		/* Drop interleaved "more-data" reports. We return "true" to
+		 * indicate that we have processed the report, but we silently
+		 * drop it.
+		 */
+		LOG_DBG("dropping interleaved fragmented report");
+		return true;
+	}
+
+	if (!reassembling && !borked) {
+		/* Try to allocate a backing buffer in which to reassemble data
+		 * from the fragmented advertising reports. If the chain is
+		 * borked, we don't bother with allocation as we will discard
+		 * the chain anyways.
+		 */
+		init_reassembly(r, &evt->addr, evt->sid);
+	}
 
-		/* No more data coming from the controller.
-		 * Create event.
+	if (r->report == NULL) {
+		LOG_DBG("No reassembly buf.");
+		start_discarding(r);
+	} else if (evt->length > net_buf_tailroom(r->report)) {
+		LOG_DBG("Adv report doesn't fit reassembly buffer.");
+		start_discarding(r);
+	}
+
+	bool discarding = r->state == FRAG_ADV_DISCARDING;
+
+	if (!discarding) {
+		__ASSERT_NO_MSG(r->report);
+		net_buf_add_mem(r->report, /* flex the member 🤖💪 */ evt->data, evt->length);
+	}
+
+	if (!discarding && complete) {
```

Comment:
I think we do need the `!discarding`. Else we will enqueue `r->report` which is `NULL` when we are discarding. We can of course rework the logic to have this condition if we really want to.

Reply:
My point here was that we already have a `!discarding` clause and the `complete` check could be moved into the existing `!discarding` check so we don't have to check that twice

Reply:
I was reading your comment from vscode and it butchered the suggestion, sorry about that.
I tend to favor less indentation if possible but won't maintain this code. So it's up to @alwa-nordic and @jhedberg to decide.

Reply:
@jori-nordic @Thalley I'm also in general in favor of less indentation, either through playing around with conditions and function return locations, or by refactoring parts of the code to helper functions.

Reply:
If we can assume that the compiler can optimize it so that

```c
if (foo) {
   bar();
}

if (foo && zoo) {
   zar();
}
```

Generates the same binary as 

```c
if (foo) {
    bare();
    
    if (zoo) {
        zar();
    }
}
```

Then I'm absolutely fine with it :) If `foo` needs to be evaluated twice for the sake of readability, then we are effectively saying that performance is less important than readability, which can be a slippery slope in some cases :)